### PR TITLE
Add router tests and fix related issues

### PR DESCRIPTION
**Pull request by buger**
Integration router tests run all combinations of using `goagain`, `config.HttpServerOptions.OverrideDefaults`, and control API on a separate port.

Following fixes are made:

* Fixed `/tyk/api` endpoint without slash #372
* Fixed control API when `config.HttpServerOptions.OverrideDefaults` is `true` (both when with
Goagain and without) #612 
* Fixed: do not bind control API if RPC_EmergencyMode turned on #612
* Fix Control API on a separate port in forked process #612
* Potential #531 fix by not using `http.Handle("/", …)` in favor of `http.Serve`

### CODE REVIEW

**Comment by buger** [comment]
@matiasinsaurralde since it touch Control API functionality, can you check this changes?

**Comment by mvdan** [comment]
You can't add `start()` to `TestMain` like that, as it also calls `reloadLoop`. If you do, you must move it out of `start`.

**Comment by mvdan** [comment]
Also, could you split up the fixes in separate commits? Otherwise, if any of the independent changes needs reverting, it would be much harder. It's also easier to see what code changes are from what fix.

**Comment by mvdan** [comment]
Oh, forgot an important point - having fixes as individual commits will also enable backporting into 2.3. At least a couple of these we want to backport, I think.

**Comment by mvdan** [comment]
I have to run now, but I'll review this tomorrow morning.

File: main.go [hunk 1]
```diff
@@ -1370,6 +1370,10 @@ func listen(l, controlListener net.Listener, err error) {
 			if specs != nil {
 				loadApps(specs, defaultRouter)
 				getPolicies()
+
+				if config.ControlAPIPort > 0 {
+					loadAPIEndpoints(controlRouter)
+				}
```
Comment (mvdan) [inline main.go]: > If it bind on separate port, it not works

"If it binds on a separate port, it doesn't work."

File: main.go [hunk 2]
```diff
@@ -1380,6 +1390,16 @@ func listen(l, controlListener net.Listener, err error) {
...
+
+			if controlListener != nil {
+				cs := &http.Server{
+					ReadTimeout:  time.Duration(readTimeout) * time.Second,
+					WriteTimeout: time.Duration(writeTimeout) * time.Second,
+					Handler:      controlRouter,
+				}
+				go cs.Serve(controlListener)
+			}
+
```
Comment (mvdan) [inline main.go]: These two blocks are identical copies. Dedup in a func?

File: gateway_test.go [hunk 3]
```diff
@@ -804,3 +805,127 @@ func TestWithAnalyticsErrorResponse(t *testing.T) {
...
+// Admin api located on separate port
+func TestControlListener(t *testing.T) {
+	tests := []tykHttpTest{
+		{"GET", "/", 404, false, false},
+		{"GET", "/tyk/apis", 404, false, false},
+
+		// Querying control API
+		{"GET", "/", 404, false, true},
+		{"GET", "/tyk/apis", 403, false, true},
+		{"GET", "/tyk/apis/", 200, true, true},
```
Comment (mvdan) [inline gateway_test.go]: My only nit with this is that you have to mentally keep track of what each bool means. Perhaps using keyed vaues would help. If you think that's overkill for just the two bools, leaving it as-is sounds fine too.

File: main.go (hunk 2 above)
Comment (buger) [inline main.go]: Yeah, we can do that in future, this block does have differences, but minor ones. Let's make it as separate PR, maybe you will touch this part someday :)

File: gateway_test.go (hunk 3 above)
Comment (buger) [inline gateway_test.go]: Yeah, I do not like this set of booleans as well. Will think about using keyed values.

**Pull request merged**